// gatsby-interface/dist/src/components/Announcement/Announcement.esm.js
import { extends as _extends, defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import background from './assets/background.esm.js';

var baseCss = function baseCss(theme) {
  return _defineProperty({
    alignItems: "center",
    background: theme.colors.teal[5],
    color: theme.colors.teal[90],
    display: "flex",
    fontSize: theme.fontSizes[1],
    fontFamily: theme.fonts.system,
    padding: "".concat(theme.space[5], " ").concat(theme.space[9]),
    backgroundImage: "url(".concat(background, ")"),
    backgroundPosition: "right center",
    backgroundRepeat: "no-repeat"
  }, "&:not(:first-child)", {
    borderTop: "1px solid ".concat(theme.colors.teal[10])
  });
};

function Announcement(props) {
  return jsx("div", _extends({
    css: baseCss
  }, props));
}
Announcement.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Announcement"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Announcement/Announcement.tsx"] = {
    name: "Announcement",
    docgenInfo: Announcement.__docgenInfo,
    path: "src/components/Announcement/Announcement.tsx"
  };
}

export { Announcement };
